src/extension.ts

```diff
@@ -110,4 +113,89 @@ export const activate = () => {
         waitUntil(performFixes())
     })
+
+    vscode.workspace.onDidChangeTextDocument(
+        ({ contentChanges, document }) => {
+            if (!getExtensionSetting("insertMissingCommaOnEnter")) {
+                return;
+            }
+
+            if (
+                document.languageId !== "json" &&
+                document.languageId !== "jsonc"
+            ) {
+                return;
+            }
+
+            if (contentChanges.length === 0) {
+                return;
+            }
+
+            const editor = vscode.window.activeTextEditor;
+
+            if (
+                document.uri !== editor?.document.uri ||
+                ["output"].includes(editor.document.uri.scheme)
+            ) {
+                return;
+            }
+            if (
+                vscode.workspace.fs.isWritableFileSystem(
+                    document.uri.scheme
+                ) === false
+            ) {
+                return;
+            }
+
+            editor.edit((edit) => {
+                contentChanges.forEach(async (content) => {
+                    if (
+                        !content.text.startsWith("\n") &&
+                        !content.text.startsWith("\r\n")
+                    ) {
+                        return;
+                    }
+
+                    const prevLinePosition = document.positionAt(content.rangeOffset);
+
+                    const prevLine = document.lineAt(prevLinePosition);
+                    const prevLineText = prevLine.text;
+
+                    const prevLineLastChar = prevLineText.at(-1);
+
+                    if (!prevLineLastChar) {
+                        return;
+                    }
+
+                    if (document.languageId === 'jsonc') {
+                        if (prevLineText.trim().startsWith("//") || prevLineText.trim().startsWith("/*")) {
+                            return;
+                        }
+
+                        const textWithouComments = stripJsonComments(document.getText());
+                        const prevLineTextWithoutComments = getTextByLine(textWithouComments, prevLine.lineNumber)?.trim();
+
+                        if (prevLineTextWithoutComments !== prevLineText.trim()) {
+                            return;
+                        }
+                    }
+
+                    const isCurrentLineEmpty =
+                        document.lineAt(prevLine.lineNumber + 1).text.trim() === "";
+
+                    const isMatchValue =
+                        prevLineLastChar === "}" ||
+                        prevLineLastChar === '"' ||
+                        prevLineLastChar === ']' ||
+                        isNumber(prevLineLastChar);
+
+                    if (isMatchValue && isCurrentLineEmpty) {
+                        // In multicursor mode last character position is broken when I use prevLinePosition, IDK why
```

Comment:
I faced similar probleim in
https://github.com/zardoy/vscode-better-snippets/blob/2ce9b402d94759cfe795d0f087dcfe5134b5523e/src/extension.ts#L391

This is because here, you work with **already changed document** (`document.lineAt`). For example you have the following document (| - cursor position):

```
1. {
2.     "test1": "a"|
3.     "test2": "a"|
4. }
```

After pressing enter:

```
1. {
2.    "test1": "a"
3.    |
4.    "test2": "a"
5.    |
6. }
```

In this callback you receive information that lines `2` and `3` has typed `\n` character. And now you can guess what line content you receive at the end (`"` and ` `). Also you were working with unsorted contentChanges so most probably code was trying first to get 3rd line and only then 2nd.

> Note: I fixed it with last commit, do you mind testing it?

Reply:
Tested, works very well, thanks for the help

![Code_6ZZi5G5vds](https://user-images.githubusercontent.com/68181944/203049286-ee7bd1ab-4d43-4788-975a-432c37a29513.gif)